README.md
# nv_templates

A "Story Templates" manager plugin for [noveltree](https://peter88213.github.io/noveltree)

Adds a **Story Templates** entry to the **Tools** menu to open a submenu for loading and saving the story structure. 

For more information, see the [project homepage](https://peter88213.github.io/nv_templates) with description and download instructions.

## Feedback? Ideas? Feature requests?

You can go to the [noveltree "discussions" forum](https://github.com/peter88213/noveltree/discussions) and start a thread.

# Contributing

## How to provide translations

First, you need to know your language code according to ISO 639-1.

For English, this is, for example, `en`, for German, it is `de`.

**NOTE:** The procedure described below is greatly simplified if you create a language pack based on the [noveltree_xx](https://github.com/peter88213/noveltree_xx) template and use the tools provided for this purpose. 

### Create a message catalog

A "message catalog" is a dictionary for the plugin's messages and menu entries.

For creating a message catalog, you download a template with all English messages from [here](https://github.com/peter88213/nv_templates/blob/main/i18n/messages.pot). 


Rename `messages.pot` to `<your language code>.po`, then give some specific information in the header data by modifying the following lines:

```
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language: LANGUAGE\n"
```

**NOTE:** Be sure to use a text editor that writes utf-8 encoded text. Otherwise, it may not work with non-ASCII characters used in your language.

The  `<your language code>.po` dictionary is organized as a set of *message ID (msgid)* - *message string (msgstr)* pairs, where *msgid* means the English term, and *msgstr* means the translated term. This is an example for such a pair where the message string is still missing:

```
msgid "Cannot overwrite file"
msgstr ""
```

Now you enter all missing message strings. 
- If a message ID contains placeholders like `{}`, be sure to put them also into the message string.  
- If a message ID starts with `!`, the message string must also start with `!`. 

Before you distribute your translations, you can convert and install the message catalog for testing. 

### Convert the message catalog to binary format

The plugin needs the message catalog in binary format. This is easily achieved using the **msgfmt.py** converter script. 
You find it in your Python installation, in the **Tools/i18n** subdirectory. If not, you can download the code from [here](https://github.com/python/cpython/blob/main/Tools/i18n/msgfmt.py)

Name the binary file **nv_templates.mo**. 


### Install your translation for testing

Add a subdirectory tree to **noveltree/locale**, and place *nv_templates.mo* there, like this:

```
<your home directory>
└── .novxlib/
    └── noveltree/
        └── locale/
            └─ <language code>/
               └─ LC_MESSAGES/
                  └─ nv_templates.mo
```

Then start *noveltree* and see whether your translation works. 

**NOTE:** At startup, *noveltree* tries to load a message dictionary that fits to the system language. If it doesn't find a matching language code in the *locale* directory, it uses English as default language. 

**HINT:** *nv_templates* comes with German translations. Look at the `de` directory tree, if you need an example. 


### Contribute your translations

If *noveltree* works fine with your translations, you can consider contributing it. 

An easy way may be to put a posting in the [noveltree forum](https://github.com/peter88213/noveltree/discussions), appending your  `<your language code>.po` file. 


## Development

### Conventions

See https://github.com/peter88213/novxlib/blob/main/docs/conventions.md

## Development tools

- [Python](https://python.org) version 3.10.
- [Eclipse IDE](https://eclipse.org) with [PyDev](https://pydev.org) and *EGit*.
- *Apache Ant* is used for building the application.

## License

This is Open Source software, and the *nv_templates* plugin is licensed under GPLv3. See the
[GNU General Public License website](https://www.gnu.org/licenses/gpl-3.0.en.html) for more
details, or consult the [LICENSE](https://github.com/peter88213/nv_templates/blob/main/LICENSE) file.


docs/changelog.md
[Project home page](index) > Changelog

------------------------------------------------------------------------

## Changelog


### v1.2.1

- Exclude chapters in the planning section from export that have no arc definition.

Compatibility: noveltree v4.12 API

### v1.2.0 

- Update the conditions for creating arc defining chapters (see instructions for use).

Compatibility: noveltree v4.12 API

### v1.1.1 

- Abort loading silently, if no template is selected.
- Update sample templates.

Compatibility: noveltree v4.12 API

### v1.1.0 

- Add "open folder" command.

Compatibility: noveltree v4.12 API

### v1.0.1 

- When reading Markdown, discard trailing spaces.
- Add sample templates: Hero's Journey, 3 Act Structure, 4 Act Structure.

Compatibility: noveltree v4.12 API

### v1.0.0 

- Define arcs as described in the help file when creating Todo chapters.

Compatibility: noveltree v4.12 API

### v0.2.1 

- Provide online help.

Compatibility: noveltree v4.12 API

### v0.2.0 

- Add "Save" command.

Compatibility: noveltree v4.12 API

### v0.1.1 

- Prevent new "Narrative" chapters from being auto-numbered on creation.

Compatibility: noveltree v4.12 API

### v0.1.0 

Proof of concept. 

Compatibility: noveltree v4.11 API


test/novelyst_templates.py
"""A "Story Templates" plugin for novelyst.

Version @release

Adds a 'Add Story Templates' entry to the 'Tools' menu to open a window
with a combobox that lists all available themes. 
The selected theme will be persistently applied.  

Requires Python 3.6+
Copyright (c) 2023 Peter Triesberger
For further information see https://github.com/peter88213/novelyst_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import tkinter as tk
from tkinter import messagebox
from tkinter import filedialog
from pathlib import Path
import sys
import gettext
import locale

__all__ = ['Error',
           '_',
           'norm_path',
           'LOCALE_PATH',
           'CURRENT_LANGUAGE',
           'APPLICATION',
           'PLUGIN',
           ]


class Error(Exception):
    pass


LOCALE_PATH = f'{os.path.dirname(sys.argv[0])}/locale/'
try:
    CURRENT_LANGUAGE = locale.getlocale()[0][:2]
except:
    CURRENT_LANGUAGE = locale.getdefaultlocale()[0][:2]
try:
    t = gettext.translation('novelyst_templates', LOCALE_PATH, languages=[CURRENT_LANGUAGE])
    _ = t.gettext
except:

    def _(message):
        return message

APPLICATION = _('Story Templates')
PLUGIN = f'{APPLICATION} plugin @release'


def norm_path(path):
    if path is None:
        path = ''
    return os.path.normpath(path)



class MdTemplate:
    DESCRIPTION = _('Story Template')
    EXTENSION = '.md'

    def __init__(self, filePath, ui):
        self.filePath = filePath
        self._ui = ui

    def read(self):
        try:
            with open(self.filePath, 'r', encoding='utf-8') as f:
                content = f.read()
        except(FileNotFoundError):
            raise Error(f'{_("File not found")}: "{norm_path(self.filePath)}".')
        except:
            raise Error(f'{_("Cannot read file")}: "{norm_path(self.filePath)}".')

        mdLines = content.split('\n')
        parent = None
        newElement = None
        desc = []
        for mdLine in mdLines:
            if mdLine.startswith('#'):
                if newElement is not None:
                    newElement.desc = '\n'.join(desc)
                    desc = []
                    newElement = None
            if mdLine.startswith('####'):
                if not parent:
                    continue

                newTitle = mdLine[5:]
                scId = self._ui.tv.add_scene(title=newTitle, scType=2)
                if scId:
                    newElement = self._ui.novel.scenes[scId]
            elif mdLine.startswith('###'):
                if not parent:
                    continue

                newTitle = mdLine[4:]
                chId = self._ui.tv.add_chapter(title=newTitle, chType=2)
                if chId:
                    newElement = self._ui.novel.chapters[chId]
            elif mdLine.startswith('##'):
                if not parent:
                    continue

                newTitle = mdLine[3:]
                chId = self._ui.tv.add_part(title=newTitle, chType=2)
                if chId:
                    newElement = self._ui.novel.chapters[chId]
            elif mdLine.startswith('#'):

                parent = mdLine[2:]
                try:
                    self._ui.tv.tree.selection_set(parent)
                except:
                    raise Error(f'{_("Wrong heading")}: "{parent}"')
            elif mdLine:
                desc.append(mdLine)

    def write(self):
        raise Error(f'{self.filePath} Write method is not implemented.')



class Plugin():
    """A 'Story Templates' plugin class.
    
    Public methods:
        disable_menu() -- disable menu entries when no project is open.
        enable_menu() -- enable menu entries when a project is open.    
    """
    VERSION = '@release'
    NOVELYST_API = '4.12'
    DESCRIPTION = 'Creates a story arc framework from a markdown template'
    URL = 'https://peter88213.github.io/novelyst_templates'

    def install(self, ui):
        """Add a submenu to the 'Tools' menu.
        
        Positional arguments:
            ui -- reference to the NovelystTk instance of the application.
        """
        self._ui = ui

        self._templatesMenu = tk.Menu(self._ui.toolsMenu, tearoff=0)
        self._templatesMenu.add_command(label=_('Load'), command=self._load_template)
        self._templatesMenu.add_command(label=_('Save'), command=self._save_template)

        self._ui.toolsMenu.add_cascade(label=APPLICATION, menu=self._templatesMenu)
        self._ui.toolsMenu.entryconfig(APPLICATION, state='disabled')
        self._fileTypes = [(MdTemplate.DESCRIPTION, MdTemplate.EXTENSION)]

    def disable_menu(self):
        self._ui.toolsMenu.entryconfig(APPLICATION, state='disabled')

    def enable_menu(self):
        self._ui.toolsMenu.entryconfig(APPLICATION, state='normal')

    def _load_template(self):
        try:
            homeDir = str(Path.home()).replace('\\', '/')
            templateDir = f'{homeDir}/.pywriter/novelyst/templates'
        except:
            templateDir = '.'
        fileName = filedialog.askopenfilename(filetypes=self._fileTypes,
                                              defaultextension=self._fileTypes[0][1],
                                              initialdir=templateDir)
        try:
            templates = MdTemplate(fileName, self._ui)
            templates.read()
        except Error as ex:
            messagebox.showerror(_('Template loading aborted'), str(ex))

    def _save_template(self):
        try:
            homeDir = str(Path.home()).replace('\\', '/')
            templateDir = f'{homeDir}/.pywriter/novelyst/templates'
        except:
            templateDir = '.'
        fileName = filedialog.asksaveasfilename(filetypes=self._fileTypes,
                                              defaultextension=self._fileTypes[0][1],
                                              initialdir=templateDir)
        if not fileName:
            return

        try:
            templates = MdTemplate(fileName, self._ui)
            templates.write()
        except Error as ex:
            messagebox.showerror(_('Cannot save template'), str(ex))


tools/make_pot.py
"""Generate a template file (pot) for message translation.

For further information see https://github.com/peter88213/noveltree_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
import build_novelyst_templates as build_plugin
sys.path.insert(0, f'{os.getcwd()}/../../novxlib/src')
import pgettext

APP = 'novelyst_templates'
POT_FILE = '../i18n/messages.pot'


def make_pot(version='unknown'):
    # Generate a complete plugin.
    build_plugin.main()

    # Generate a pot file from the script.
    if os.path.isfile(POT_FILE):
        os.replace(POT_FILE, f'{POT_FILE}.bak')
        backedUp = True
    else:
        backedUp = False
    try:
        pot = pgettext.PotFile(POT_FILE, app=APP, appVersion=version)
        pot.scan_file(build_plugin.TARGET_FILE)
        print(f'Writing "{pot.filePath}"...\n')
        pot.write_pot()

    except:
        if backedUp:
            os.replace(f'{POT_FILE}.bak', POT_FILE)
        print('WARNING: Cannot write pot file.')
        return False

    return True


if __name__ == '__main__':
    try:
        success = make_pot(sys.argv[1])
    except:
        success = make_pot()
    if not success:
        sys.exit(1)


tools/translate_de.py
"""Generate German translation files for GNU gettext.

- Update the project's 'de.po' translation file.
- Generate the language specific 'novxlib.mo' dictionary.

Usage: 
translate_de.py

File structure:

├── novxlib/
│   ├── i18n/
│   │   └── de.json
│   └── src/
│       ├── translations.py
│       └── msgfmt.py
└── nv_templates/
    ├── src/ 
    ├── tools/ 
    │   └── translate_de.py
    └── i18n/
        ├── messages.pot
        ├── de.po
        └── locale/
            └─ de/
               └─ LC_MESSAGES/
                  └─ nv_templates.mo
    
Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import translations
from shutil import copyfile
import msgfmt

APP_NAME = 'nv_templates'
PO_PATH = '../i18n/de.po'
MO_PATH = f'../i18n/locale/de/LC_MESSAGES/{APP_NAME}.mo'
MO_COPY = f'../../kalliope/src/locale/de/LC_MESSAGES/{APP_NAME}.mo'


def main(version='unknown'):
    if translations.main('de', app=APP_NAME, appVersion=version, json=True):
        print(f'Writing "{MO_PATH}" ...')
        msgfmt.make(PO_PATH, MO_PATH)
        copyfile(MO_PATH, MO_COPY)
    else:
        sys.exit(1)


if __name__ == '__main__':
    try:
        main(sys.argv[1])
    except:
        main()


src/novelyst_templates.py
"""A "Story Templates" plugin for novelyst.

Version @release

Adds a 'Add Story Templates' entry to the 'Tools' menu to open a window
with a combobox that lists all available themes. 
The selected theme will be persistently applied.  

Requires Python 3.6+
Copyright (c) 2023 Peter Triesberger
For further information see https://github.com/peter88213/novelyst_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
"""
import os
import tkinter as tk
import webbrowser
from tkinter import messagebox
from tkinter import filedialog
from pathlib import Path
from nvtemplateslib.nvtemplates_globals import *
from nvtemplateslib.md_template import MdTemplate


class Plugin():
    """A 'Story Templates' plugin class.
    
    Public methods:
        disable_menu() -- disable menu entries when no project is open.
        enable_menu() -- enable menu entries when a project is open.    
    """
    VERSION = '@release'
    NOVELYST_API = '4.45'
    DESCRIPTION = 'A "Story Templates" manager'
    URL = 'https://peter88213.github.io/novelyst_templates'
    _HELP_URL = 'https://peter88213.github.io/novelyst_templates/usage'

    def install(self, ui):
        """Add a submenu to the 'Tools' menu.
        
        Positional arguments:
            ui -- reference to the NovelystTk instance of the application.
        """
        self._ui = ui
        try:
            homeDir = str(Path.home()).replace('\\', '/')
            self._templateDir = f'{homeDir}/.pywriter/novelyst/templates'
        except:
            self._templateDir = '.'

        # Create "Story Templates" submenu.
        self._templatesMenu = tk.Menu(self._ui.toolsMenu, tearoff=0)
        self._templatesMenu.add_command(label=_('Load'), command=self._load_template)
        self._templatesMenu.add_command(label=_('Save'), command=self._save_template)
        self._templatesMenu.add_command(label=_('Open folder'), command=self._open_folder)

        # Create Tools menu entry.
        self._ui.toolsMenu.add_cascade(label=APPLICATION, menu=self._templatesMenu)
        self._ui.toolsMenu.entryconfig(APPLICATION, state='disabled')
        self._fileTypes = [(MdTemplate.DESCRIPTION, MdTemplate.EXTENSION)]

        # Add an entry to the Help menu.
        self._ui.helpMenu.add_command(label=_('Templates plugin Online help'), command=lambda: webbrowser.open(self._HELP_URL))

    def disable_menu(self):
        """Disable menu entries when no project is open."""
        self._ui.toolsMenu.entryconfig(APPLICATION, state='disabled')

    def enable_menu(self):
        """Enable menu entries when a project is open."""
        self._ui.toolsMenu.entryconfig(APPLICATION, state='normal')

    def _load_template(self):
        """Create a structure of "Todo" chapters and scenes from a Markdown file."""
        fileName = filedialog.askopenfilename(filetypes=self._fileTypes,
                                              defaultextension=self._fileTypes[0][1],
                                              initialdir=self._templateDir)
        if fileName:
            try:
                templates = MdTemplate(fileName, self._ui)
                templates.read()
            except Error as ex:
                messagebox.showerror(_('Template loading aborted'), str(ex))

    def _save_template(self):
        """Save a structure of "Todo" chapters and scenes to a Markdown file."""
        fileName = filedialog.asksaveasfilename(filetypes=self._fileTypes,
                                              defaultextension=self._fileTypes[0][1],
                                              initialdir=self._templateDir)
        if not fileName:
            return

        try:
            templates = MdTemplate(fileName, self._ui)
            templates.write()
        except Error as ex:
            messagebox.showerror(_('Cannot save template'), str(ex))

        self._ui.set_info_how(_('Template saved.'))

    def _open_folder(self):
        """Open the templates folder with the OS file manager."""
        try:
            os.startfile(norm_path(self._templateDir))
            # Windows
        except:
            try:
                os.system('xdg-open "%s"' % norm_path(self._templateDir))
                # Linux
            except:
                try:
                    os.system('open "%s"' % norm_path(self._templateDir))
                    # Mac
                except:
                    pass


tools/build_nv_templates.py
"""Build a templates noveltree plugin.
        
In order to distribute a single script without dependencies, 
this script "inlines" all modules imported from the novxlib package.

The novxlib project (see https://github.com/peter88213/novxlib)
must be located on the same directory level as the nv_templates project. 

Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import inliner

SRC = '../src/'
BUILD = '../test/'
SOURCE_FILE = f'{SRC}nv_templates.py'
TARGET_FILE = f'{BUILD}nv_templates.py'


def main():
    inliner.run(SOURCE_FILE, TARGET_FILE, 'nvtemplateslib', '../../novelyst_templates/src/')
    inliner.run(TARGET_FILE, TARGET_FILE, 'noveltreelib', '../../noveltree/src/')
    print('Done.')


if __name__ == '__main__':
    main()


src/nvtemplateslib/md_template.py
"""Provide a class for Markdown story template representation.

Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
from nvtemplateslib.nvtemplates_globals import *


class MdTemplate:
    """Markdown story template representation.
    
    Public instance variables:
        filePath: str -- path to the file (property with getter and setter). 
    """
    DESCRIPTION = _('Story Template')
    EXTENSION = '.md'
    ARC_MARKER = '-'

    def __init__(self, filePath, controller, ui):
        """Initialize instance variables.

        Positional arguments:
            filePath: str -- path to the file represented by the File instance.
        """
        self.filePath = filePath
        self._controller = controller
        self._ui = ui

    def read(self):
        """Parse the Markdown file and create parts, chapters, and sections.
        
        Raise the "Error" exception in case of error. 
        """
        try:
            with open(self.filePath, 'r', encoding='utf-8') as f:
                mdLines = f.readlines()
        except(FileNotFoundError):
            raise Error(f'{_("File not found")}: "{norm_path(self.filePath)}".')
        except:
            raise Error(f'{_("Cannot read file")}: "{norm_path(self.filePath)}".')

        if mdLines[0].startswith('# nv'):
            self.read_novelyst_structure(mdLines)
        else:
            self.read_noveltree_structure(mdLines)

    def read_novelyst_structure(self, mdLines):
        chId = self._controller.c_add_chapter(targetNode=CH_ROOT, title=_('Stages'), chLevel=2, chType=3)
        scId = chId
        arcSection = False
        newElement = None
        desc = []
        for mdLine in mdLines:
            mdLine = mdLine.strip()
            if mdLine.startswith('#'):
                if newElement is not None:
                    newElement.desc = ''.join(desc).strip().replace('  ', ' ')
                    desc = []
                    newElement = None
                if mdLine.startswith('####'):
                    if arcSection:
                        # Add a second-level stage.
                        newTitle = mdLine[5:]
                        scId = self._controller.c_add_stage(targetNode=scId, title=newTitle, scType=3)
                        if scId:
                            newElement = self._controller.novel.sections[scId]
                elif mdLine.startswith('###'):
                    if not arcSection:
                        # Add a first-level stage.
                        newTitle = mdLine[4:]
                        scId = self._controller.c_add_stage(targetNode=scId, title=newTitle, scType=2)
                        if scId:
                            newElement = self._controller.novel.sections[scId]
                elif mdLine.strip() == '# pl':
                    arcSection = True
            elif mdLine:
                desc.append(f'{mdLine} ')
            else:
                desc.append('\n')
        try:
            newElement.desc = ''.join(desc).strip().replace('  ', ' ')
        except AttributeError:
            pass

    def read_noveltree_structure(self, mdLines):
        if self._controller.novel.chapters:
            self.list_stages(mdLines)
        else:
            self.create_chapter_structure(mdLines)

    def create_chapter_structure(self, mdLines):
        i = 0
        chId = CH_ROOT
        addChapter = True
        newElement = None
        desc = []
        for mdLine in mdLines:
            mdLine = mdLine.strip()
            if mdLine.startswith('#'):
                if newElement is not None:
                    newElement.desc = ''.join(desc).strip().replace('  ', ' ')
                    desc = []
                    newElement = None
                if mdLine.startswith('## '):
                    # Add a 2nd level stage.
                    if addChapter:
                        i += 1
                        chId = self._controller.c_add_chapter(targetNode=chId, title=f"{_('Chapter')} {i}", chLevel=2, chType=0)
                        scId = chId
                    newTitle = mdLine[3:].strip()
                    scId = self._controller.c_add_stage(targetNode=scId, title=newTitle, scType=3)
                    newElement = self._controller.novel.sections[scId]
                    scId = self._controller.c_add_section(targetNode=scId, title=_('New Section'), scType=0, status=1)
                    addChapter = True
                elif mdLine.startswith('# '):
                    # Add a ist level stage.
                    i += 1
                    chId = self._controller.c_add_chapter(targetNode=chId, title=f"{_('Chapter')} {i}", chLevel=2, chType=0)
                    newTitle = mdLine[2:].strip()
                    scId = self._controller.c_add_stage(targetNode=chId, title=newTitle, stageLevel=1)
                    newElement = self._controller.novel.sections[scId]
                    addChapter = False
                else:
                    scId = None
            elif mdLine:
                desc.append(f'{mdLine} ')
            else:
                desc.append('\n')
        newElement.desc = ''.join(desc).strip().replace('  ', ' ')

    def list_stages(self, mdLines):
        chId = self._controller.c_add_chapter(targetNode=CH_ROOT, title=_('Stages'), chLevel=2, chType=3)
        scId = chId
        newElement = None
        desc = []
        for mdLine in mdLines:
            mdLine = mdLine.strip()
            if mdLine.startswith('#'):
                if newElement is not None:
                    newElement.desc = ''.join(desc).strip().replace('  ', ' ')
                    desc = []
                    newElement = None
                if mdLine.startswith('## '):
                    # Add a 2nd level stage.
                    newTitle = mdLine[3:].strip()
                    scId = self._controller.c_add_stage(targetNode=scId, title=newTitle, scType=3)
                elif mdLine.startswith('# '):
                    # Add a 1st level stage.
                    newTitle = mdLine[2:].strip()
                    scId = self._controller.c_add_stage(targetNode=scId, title=newTitle, scType=2)
                else:
                    scId = None
                if scId:
                    newElement = self._controller.novel.sections[scId]
            elif mdLine:
                desc.append(f'{mdLine} ')
            else:
                desc.append('\n')
        newElement.desc = ''.join(desc).strip().replace('  ', ' ')

    def write(self):
        """Iterate the project structure and write the new elements to a Markdown file.
        
        Raise the "Error" exception in case of error. 
        """
        mdLines = []
        for chId in self._controller.novel.tree.get_children(CH_ROOT):
            for scId in self._controller.novel.tree.get_children(chId):
                if self._controller.novel.sections[scId].scType == 2:
                    mdLines.append(f'# {self._controller.novel.sections[scId].title}')
                    desc = self._controller.novel.sections[scId].desc
                    if desc:
                        mdLines.append(desc.replace('\n', '\n\n'))
                elif self._controller.novel.sections[scId].scType == 3:
                    mdLines.append(f'## {self._controller.novel.sections[scId].title}')
                    desc = self._controller.novel.sections[scId].desc
                    if desc:
                        mdLines.append(desc.replace('\n', '\n\n'))

        content = '\n\n'.join(mdLines)
        try:
            with open(self.filePath, 'w', encoding='utf-8') as f:
                f.write(content)
        except:
            raise Error(f'{_("Cannot write file")}: "{norm_path(self.filePath)}".')



tools/build_novelyst_templates.py
"""Build a templates novelyst plugin.
        
In order to distribute a single script without dependencies, 
this script "inlines" all modules imported from the novxlib package.

The novxlib project (see https://github.com/peter88213/novxlib)
must be located on the same directory level as the novelyst_templates project. 

Copyright (c) 2023 Peter Triesberger
For further information see https://github.com/peter88213/noveltree_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib/src')
import inliner

SRC = '../src/'
BUILD = '../test/'
SOURCE_FILE = f'{SRC}novelyst_templates.py'
TARGET_FILE = f'{BUILD}novelyst_templates.py'


def main():
    inliner.run(SOURCE_FILE, TARGET_FILE, 'nvtemplateslib', '../../novelyst_templates/src/')
    inliner.run(TARGET_FILE, TARGET_FILE, 'novelystlib', '../../novelyst/src/')
    print('Done.')


if __name__ == '__main__':
    main()


src/nv_templates.py
"""A "Story Templates" plugin for noveltree.

Version @release

Adds a 'Add Story Templates' entry to the 'Tools' menu to open a window
with a combobox that lists all available themes. 
The selected theme will be persistently applied.  

Requires Python 3.6+
Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_templates
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
"""
import os
import tkinter as tk
import webbrowser
from tkinter import messagebox
from tkinter import filedialog
from pathlib import Path
from nvtemplateslib.nvtemplates_globals import *
from nvtemplateslib.md_template import MdTemplate


class Plugin():
    """A 'Story Templates' plugin class."""
    VERSION = '@release'
    NOVELYST_API = '0.6'
    DESCRIPTION = 'A "Story Templates" manager'
    URL = 'https://peter88213.github.io/nv_templates'
    _HELP_URL = 'https://peter88213.github.io/nv_templates/usage'

    def install(self, controller, ui):
        """Add a submenu to the 'Tools' menu.
        
        Positional arguments:
            controller -- reference to the main controller instance of the application.
            ui -- reference to the main view instance of the application.
        """
        self._ui = ui
        self._controller = controller
        try:
            homeDir = str(Path.home()).replace('\\', '/')
            self._templateDir = f'{homeDir}/.noveltree/templates'
        except:
            self._templateDir = '.'

        # Create "Story Templates" submenu.
        self._templatesMenu = tk.Menu(self._ui.toolsMenu, tearoff=0)
        self._templatesMenu.add_command(label=_('Load'), command=self._load_template)
        self._templatesMenu.add_command(label=_('Save'), command=self._save_template)
        self._templatesMenu.add_command(label=_('Open folder'), command=self._open_folder)

        # Add an entry to the "File > New" menu.
        self._ui.newMenu.add_command(label=_('Create from template...'), command=self._new_project)

        # Create Tools menu entry.
        self._ui.toolsMenu.add_cascade(label=APPLICATION, menu=self._templatesMenu)
        self._ui.toolsMenu.entryconfig(APPLICATION, state='disabled')
        self._fileTypes = [(MdTemplate.DESCRIPTION, MdTemplate.EXTENSION)]

        # Add an entry to the Help menu.
        self._ui.helpMenu.add_command(label=_('Templates plugin Online help'), command=lambda: webbrowser.open(self._HELP_URL))

    def disable_menu(self):
        """Disable menu entries when no project is open."""
        self._ui.toolsMenu.entryconfig(APPLICATION, state='disabled')

    def enable_menu(self):
        """Enable menu entries when a project is open."""
        self._ui.toolsMenu.entryconfig(APPLICATION, state='normal')

    def _load_template(self):
        """Create a structure of "Todo" chapters and scenes from a Markdown file."""
        fileName = filedialog.askopenfilename(filetypes=self._fileTypes,
                                              defaultextension=self._fileTypes[0][1],
                                              initialdir=self._templateDir)
        if fileName:
            try:
                templates = MdTemplate(fileName, self._controller, self._ui)
                templates.read()
            except Error as ex:
                messagebox.showerror(_('Template loading aborted'), str(ex))

    def _new_project(self):
        """Create a noveltree project instance."""
        self._controller.new_project()
        self._load_template()

    def _open_folder(self):
        """Open the templates folder with the OS file manager."""
        try:
            os.startfile(norm_path(self._templateDir))
            # Windows
        except:
            try:
                os.system('xdg-open "%s"' % norm_path(self._templateDir))
                # Linux
            except:
                try:
                    os.system('open "%s"' % norm_path(self._templateDir))
                    # Mac
                except:
                    pass

    def _save_template(self):
        """Save a structure of "Todo" chapters and scenes to a Markdown file."""
        fileName = filedialog.asksaveasfilename(filetypes=self._fileTypes,
                                              defaultextension=self._fileTypes[0][1],
                                              initialdir=self._templateDir)
        if not fileName:
            return

        try:
            templates = MdTemplate(fileName, self._ui)
            templates.write()
        except Error as ex:
            messagebox.showerror(_('Cannot save template'), str(ex))

        self._ui.set_info_how(_('Template saved.'))

